### ISSUE TITLE
ORA-00920: invalid relational operator with CPK 12.0.0 against Oracle database

### ISSUE THREAD

**Issue by awaltman** [composite-primary-keys/composite_primary_keys#497]
I cloned the CPK repository, configured the test connection for my Oracle database and ran the unit tests. There were many errors like the following:
```
TestCreate#test_has_many_ids_1:
ActiveRecord::StatementInvalid: OCIError: ORA-00920: invalid relational operator
    stmt.c:267:in oci8lib_260.so
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/ruby-oci8-2.2.7/lib/oci8/cursor.rb:137:in `exec'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-oracle_enhanced-adapter-6.0.0/lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:146:in `exec_update'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-oracle_enhanced-adapter-6.0.0/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:152:in `block in exec_update'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-6.0.0/lib/active_record/connection_adapters/abstract_adapter.rb:707:in `block (2 levels) in log'
    /home/awaltman/.rvm/rubies/ruby-2.6.4/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-6.0.0/lib/active_record/connection_adapters/abstract_adapter.rb:706:in `block in log'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activesupport-6.0.0/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-6.0.0/lib/active_record/connection_adapters/abstract_adapter.rb:697:in `log'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-oracle_enhanced-adapter-6.0.0/lib/active_record/connection_adapters/oracle_enhanced/dbms_output.rb:36:in `log'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-oracle_enhanced-adapter-6.0.0/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:136:in `exec_update'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-6.0.0/lib/active_record/connection_adapters/abstract/database_statements.rb:176:in `delete'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-6.0.0/lib/active_record/connection_adapters/abstract/query_cache.rb:22:in `delete'
    /home/awaltman/tmp/composite_primary_keys/lib/composite_primary_keys/relation.rb:98:in `delete_all'
    /home/awaltman/.rvm/gems/ruby-2.6.4/gems/activerecord-6.0.0/lib/active_record/querying.rb:21:in `delete_all'
    /home/awaltman/tmp/composite_primary_keys/test/test_create.rb:131:in `test_has_many_ids_1'
```
I think this is because the method subquery_for in CPK 12.0.0 lib/composite_primary_keys/relation.rb specifies an alias for the subquery that begins with the underscore character: "__active_record_temp", which is not allowed in Oracle (see: https://docs.oracle.com/en/database/oracle/oracle-database/12.2/sqlrf/Database-Object-Names-and-Qualifiers.html#GUID-75337742-67FD-4EC0-985F-741C93D918DA, rule 6). I noted also that this implementation of subquery_for looks like a copy of the method from CPK 11.2.0 lib/composite_primary_keys/connection_adapters/abstract_mysql_adapter.rb and I wonder if it is MySQL specific. I modified subquery_for locally to use the alias name "active_record_temp", rebuilt my test schema and the CPK unit tests pass. However, this change does not allow the unit tests for my application to pass as there are many errors:
```
ActiveRecord::StatementInvalid: OCIError: ORA-00933: SQL command not properly ended
```
when doing a delete_all. I'll look into this a bit more.

Thanks,
Andrew

**Comment by cfis** [comment]
Hey Andrew - sorry missed your post. Did you have a chance to look into this a bit more?  Happy to take a patch to fix that name.

**Comment by codeodor** [comment]
There was one place I remember fixing to use the subquery that was mysql-specific, but worked in all the DBs I could test against. I would keep it as a subquery so we don't add more code to maintain, if possible.

**Comment by awaltman** [comment]
Thanks for reminding me! :-)

**Comment by cfis** [comment]
If those are the only changes you have, then you can push them to github and then click the "pull request" button.  Just make sure to make a pull request against the correct composite keys branch.

If you have a bunch of other changes, then its best to create a new git branch locally.  Then check in just those 2 changes to that branch and push that up to github and make the pull request.

**Comment by cfis** [comment]
So what did we decide about the chagnes to relation.rb? Specifically:

   - stmt.take(arel.limit)
   -  stmt.offset(arel.offset)
   -  stmt.order(*arel.orders)

**Comment by awaltman** [comment]
I think we should exclude the removal of those lines (and the changes for the unit test) from the fix for this issue while I investigate if the delete_all in AR 6 works if there are order by, limit and offset options on relations when using an Oracle database. 

Would you like me to resolve the previous pull request and submit a new one that just includes the renaming of the sub-query alias?

**Comment by cfis** [comment]
Yes that would be great.

**Comment by codeodor** [comment]
Just a note here. I'm not sure if it's related but I've noticed very slow queries in `update_all` now (on MySQL). I think if / when this gets tackled, it will be good to see if we can also keep the existing `wheres` as part of the update. Right now it looks up all of the records as part of a subselect, and tries to update that way, but is that how Rails does it? I wonder what would happen if we just left the query alone and let it update all based on the conditions set in the `where` clauses.

**Comment by codeodor** [comment]
Following up on my comment, removing update_all and delete_all still works with MySQL, but not PostgreSQL. Furthermore, it doesn't seem related to this Oracle issue, except incidentally since it is in a similar area of the code.

**Comment by kingdonb** [comment]
So, in regards to #503, the fixes in #499 do not suffer from the issues that the latest release of CPK for Rails 5.2 (v11.2.0) suffers from on Rails 5.2.4 and 5.2.4.1.

My test run shows only one failure, compared to 13 or 7 on each in-between version.

```
Finished in 16.251500s, 11.7528 runs/s, 30.4587 assertions/s.

  1) Failure:
TestAssociations#test_limitable_reflections [/Users/kbarret8/Desktop/devel/ruby/composite_primary_keys/test/test_associations.rb:336]:
Expected: [1, 1]
  Actual: [3, 2]

191 runs, 495 assertions, 1 failures, 0 errors, 0 skips
```

I have been working on upgrading my APIs which interface with an Oracle database and some have Composite Primary Keys, they are all on older versions of Rails and we have not upgraded many of them in a while.  I just wanted to mention in case anyone is in a similar position that I applied the changes in #499 and it solved the broken tests for me.

I may abandon my efforts to upgrade to Rails 5.2.4.1 in favor of getting directly on the latest Rails 6.0.2.1 instead, as I'm not understanding what's wrong but can definitely see the 5.2.4 issue affects my project, and will prevent me from using that version.  If the issues with merging #499 are related to problems that it creates for MySQL and not related to Oracle, that's helpful and useful to know, as I think that would mean I can just go ahead and fork it for my upgrades.

I'd like to help get this merged, though!  Thanks all involved for the work you've put in thus far.

**Comment by maurercw** [comment]
I'm encountering this same error on 12.0.2.  Any progress on getting this fixed?

**Comment by cfis** [comment]
This is fixed in 1e238902e5a5df97eb344926869d5af6ccda3787.
